.grit/patterns/openai.md

```diff
@@ -0,0 +1,236 @@
+    $has_openai_import = `false`,
+    $has_partial_import = `false`,
+    $has_sync = `false`,
+    $has_async = `false`,
+
+    // Remap errors
+    $body <: maybe contains `openai.error.$exp` => `openai.$exp` where {
+        $need_openai_import = `true`,
+    },
+
+    // Mark all the places where we they configure openai as something that requires manual intervention
+    $body <: maybe contains bubble `openai.$field = $val` as $stmt where {
+        $need_openai_import = `true`,
+        $stmt => `$stmt # TODO: Manual intervention required. This config needs to be set on the instances of OpenAI/AsyncOpenAI`
+    },
+
+    $body <: maybe contains `import openai` as $import_stmt where {
+        $has_openai_import = `true`,
+        $body <: contains bubble($has_sync, $has_async) `openai.$res.$func($params)` as $stmt where {
+            $res <: rename_resource(),
+            $func <: rename_func($has_sync, $has_async, $res, $stmt, $params),
+        },
+    },
+
+    $body <: maybe contains `from openai import $resources` as $partial_import_stmt where {
+        $has_partial_import = `true`,
+        $body <: contains bubble($has_sync, $has_async, $resources) `$res.$func($params)` as $stmt where {
+            $resources <: contains $res,
+            $res <: rename_resource(),
+            $func <: rename_func($has_sync, $has_async, $res, $stmt, $params),
+        }
+    },
+
+    if ($has_openai_import <: `true`) {
+        $import_stmt <: change_import($has_sync, $has_async, $need_openai_import),
+        if ($has_partial_import <: `true`) {
+            $partial_import_stmt => .,
+        },
+    } else if ($has_partial_import <: `true`) {
+        $partial_import_stmt <: change_import($has_sync, $has_async, $need_openai_import),
+    },
+}
+```
+
+# Change openai import to Sync
+
+```python
+import openai
+
+completion = openai.Completion.create(model="davinci-002", prompt="Hello world")
+chat_completion = openai.ChatCompletion.create(model="gpt-3.5-turbo", messages=[{"role": "user", "content": "Hello world"}])
+```
+
+```python
+from openai import OpenAI
+
+client = OpenAI()
+
+completion = client.completions.create(model="davinci-002", prompt="Hello world")
+chat_completion = client.chat.completions.create(model="gpt-3.5-turbo", messages=[{"role": "user", "content": "Hello world"}])
+```
+
+# Change openai import to Async
+
+```python
+import openai
+
+completion = await openai.Completion.acreate(model="davinci-002", prompt="Hello world")
+chat_completion = await openai.ChatCompletion.acreate(model="gpt-3.5-turbo", messages=[{"role": "user", "content": "Hello world"}])
+```
+
+```python
+from openai import AsyncOpenAI
+
+aclient = AsyncOpenAI()
+
+completion = await aclient.completions.create(model="davinci-002", prompt="Hello world")
+chat_completion = await aclient.chat.completions.create(model="gpt-3.5-turbo", messages=[{"role": "user", "content": "Hello world"}])
+```
+
+# Change openai import to Both
```

Comment:
Let's also add a case verifying that multiple calls to each API (ex. two Completions.create) reuse the same client (ie. we don't create a new client for each instance).